File: frontend/src/api/summary.ts
import { api } from './client'

export interface HourlyPoint {
  hour: string
  units: number
}

export interface ResinPoint {
  resin: string
  units: number
  litres: number
}

export interface CartridgePoint {
  cartridge_type: string
  units: number
  litres: number
}

export interface ShiftSummary {
  has_logs_today: boolean
  units: number
  litres: number
  scrap: number
  yield_pct: number
  logs_submitted: number
  has_output_logs: boolean
  hourly_timeline: HourlyPoint[]
  by_resin: ResinPoint[]
  by_cartridge: CartridgePoint[]
}

export interface MonthlyRecap {
  has_data: boolean
  month_label: string
  units: number
  best_day_ordinal: string | null
  best_day_units: number
  mismatches: number
}

export interface MilestoneTier {
  at: number
  label: string
  emoji: string
}

export interface Career {
  operator_name: string
  units_lifetime: number
  units_today: number
  current: MilestoneTier | null
  next: MilestoneTier | null
  /** How far through the gap between the tier held and the next one. */
  pct: number
  remaining: number
  tiers: MilestoneTier[]
}

export interface ShiftRecap {
  operator_name: string
  units: number
  litres: number
  scrap: number
  yield_pct: number
  logs: number
  hours_active: number
  best_hour_units: number
  rank: number | null
  ranked_of: number
  weights_taken: number
  weights_in_band: number
  badges_today: MilestoneTier[]
  lifetime: number
  next_badge: MilestoneTier | null
  to_next: number
}

export const summaryApi = {
  shiftRecap: (asOperator?: string) =>
    api.get<ShiftRecap>(asOperator ? `/summary/shift-recap?${new URLSearchParams({ as_operator: asOperator })}` : '/summary/shift-recap'),
  career: (asOperator?: string) =>
    api.get<Career>(asOperator ? `/summary/career?${new URLSearchParams({ as_operator: asOperator })}` : '/summary/career'),
  today: (asOperator?: string) =>
    api.get<ShiftSummary>(asOperator ? `/summary/today?${new URLSearchParams({ as_operator: asOperator })}` : '/summary/today'),
  monthly: () => api.get<MonthlyRecap>('/summary/monthly'),
}

File: api/schemas/summary.py
from pydantic import BaseModel


class HourlyPoint(BaseModel):
    hour: str  # ISO 8601, floored to the hour
    units: int


class ResinPoint(BaseModel):
    resin: str
    units: int
    litres: float = 0.0


class CartridgePoint(BaseModel):
    cartridge_type: str
    units: int
    litres: float = 0.0


class ShiftSummaryOut(BaseModel):
    has_logs_today: bool
    units: int
    litres: float = 0.0
    scrap: int
    yield_pct: float
    logs_submitted: int
    has_output_logs: bool  # False when today's only logs are e.g. downtime/audit
    hourly_timeline: list[HourlyPoint] = []
    by_resin: list[ResinPoint] = []
    by_cartridge: list[CartridgePoint] = []


class MilestoneTier(BaseModel):
    at: int
    label: str
    emoji: str


class CareerOut(BaseModel):
    """What somebody has poured since they started here, and the ladder of
    milestones that total sits on (see milestones.py)."""
    operator_name: str
    units_lifetime: int
    units_today: int
    current: MilestoneTier | None
    next: MilestoneTier | None
    pct: float
    remaining: int
    tiers: list[MilestoneTier]


class MonthlyRecapOut(BaseModel):
    has_data: bool
    month_label: str  # e.g. "September"
    units: int
    best_day_ordinal: str | None = None  # e.g. "12th", formatted server-side
    best_day_units: int = 0
    mismatches: int = 0


class ShiftRecapOut(BaseModel):
    """The card shown when an operator signs out or ends their shift."""
    operator_name: str
    units: int
    litres: float = 0.0
    scrap: int = 0
    yield_pct: float = 100.0
    logs: int = 0
    hours_active: float = 0.0
    best_hour_units: int = 0
    rank: int | None = None          # 1 = most units today, among people doing the same job
    ranked_of: int = 0
    weights_taken: int = 0
    weights_in_band: int = 0
    badges_today: list[MilestoneTier] = []
    lifetime: int = 0
    next_badge: MilestoneTier | None = None
    to_next: int = 0
